osu.Game/Screens/Menu/ButtonSystem.cs

```diff
@@ -81,10 +81,66 @@ public void SetOsuLogo(OsuLogo logo)
         private readonly ButtonArea buttonArea;
 
         private readonly MainMenuButton backButton;
+        private readonly MainMenuButton settingsButton;
+
+        // public for tests
+        [CanBeNull]
+        public MainMenuButton PlayButton { get; private set; }
 
         private readonly List<MainMenuButton> buttonsTopLevel = new List<MainMenuButton>();
         private readonly List<MainMenuButton> buttonsPlay = new List<MainMenuButton>();
 
+        [CanBeNull]
+        public List<MainMenuButton> CurrentButtonsList => State switch
+        {
+            ButtonSystemState.TopLevel => buttonsTopLevel,
+            ButtonSystemState.Play => buttonsPlay,
+            _ => null
+        };
+
+        private int selectionIndex = -1;
+
+        public int SelectionIndex
+        {
+            get => selectionIndex;
+            set
+            {
+                MainMenuButton previousButton = null;
+                if (CurrentSelection is MainMenuButton button)
+                    previousButton = button;
+
+                selectionIndex = value;
+
+                if (CurrentSelection is MainMenuButton newButton)
+                {
+                    if (previousButton != newButton)
+                    {
+                        previousButton?.SimulateHoverLost();
+                        newButton.SimulateHover();
+                    }
+                }
+                else
+                    previousButton?.SimulateHoverLost();
+            }
+        }
+
+        /// <summary>
+        /// The current keyboard-focused MainMenuButton, or the osu logo if
+        /// nothing valid is focused.
+        /// </summary>
+        [CanBeNull]
+        public Container CurrentSelection =>
+            (SelectionIndex >= 0 && SelectionIndex < CurrentButtonsList?.Count)
+                ? CurrentButtonsList[SelectionIndex]
+                : logo;
+
+        /// <summary>
+        /// When using keyboard navigation, mouse position at last keyboard
+        /// input. Reset to Vector2.Zero when hovering over buttons. When a
+        /// hover is registered very close to this position, we ignore the hover.
+        /// </summary>
+        private Vector2 blockedMousePosition = Vector2.Zero;
```

Comment:
Please use `Vector2?` and `null` as the special "hovering over buttons" value instead. Inline "guard" values like this `Vector2.Zero` that could be actual valid values are evil.